### PR TITLE
chore: Security Vulnerability Patch

### PR DESCRIPTION
**Pull request by uShifty**
Adds a distance check to banks to ensure nobody is executing an event when they should not be. This will also create a log displaying the nearest bank to their current position.

### CODE REVIEW

File: server/main.lua [hunk 1]
```diff
@@ -28,6 +28,23 @@ CreateThread(function()
...
+        nearestDistance = nil
+    } 
+    for k=1, #Config.BankLocations do 
+        local distance = #(coords-Config.BankLocations[k])
+        if distance<5 then 
+            returnData.isNearBank = true
+            returnData.nearestDistance = distance
+            break
+        end
+        if not returnData.nearestDistance then returnData.nearestDistance = distance elseif distance < returnData.nearestDistance then returnData.nearestDistance = distance end
```
Comment (BerkieBb) [inline server/main.lua:43]: Could you expand this if statement, it's not very readable for users now

File: server/main.lua [hunk 2]
```diff
@@ -28,6 +28,23 @@ CreateThread(function()
     end
 end)
 
+local validatePosition = function(coords)
```
Comment (BerkieBb) [inline server/main.lua:31]:
```suggestion
local function validatePosition(coords)
```
This the syntax we use everywhere

File: server/main.lua [hunk 3]
```diff
@@ -71,14 +88,17 @@ RegisterNetEvent('qb-banking:createNewCard', function()
     local src = source
     local xPlayer = QBCore.Functions.GetPlayer(src)
```
Comment (BerkieBb) [inline server/main.lua:90]:
```suggestion
if not xPlayer then return end
```
A nil check for rare-cases where a cheater or a person is not registered in qb-core either through a bug or a weird cheat

Comment (BerkieBb) [reply server/main.lua:90]: If you could, please add this check to every event there is

File: server/main.lua [hunk 4]
```diff
@@ -232,44 +252,56 @@ end)
...
-    TriggerClientEvent('qb-banking:openBankScreen', src)
-    TriggerClientEvent('QBCore:Notify', src, Lang:t('success.debit_card'), 'success')
-
-    TriggerEvent('qb-log:server:CreateLog', 'banking', 'Banking', 'lightgreen', "**"..GetPlayerName(xPlayer.PlayerData.source) .. " (citizenid: "..xPlayer.PlayerData.citizenid.." | id: "..xPlayer.PlayerData.source..")** successfully ordered a debit card")
 end)
 
 RegisterNetEvent('qb-banking:doQuickDeposit', function(amount)
     local src = source
     local xPlayer = QBCore.Functions.GetPlayer(src)
     while xPlayer == nil do Wait(0) end
```
Comment (BerkieBb) [inline server/main.lua:288]: Remove this while loop and make the if statement check which I added in the suggestion above

File: server/main.lua [hunk 5]
```diff
@@ -278,23 +310,36 @@ RegisterNetEvent('qb-banking:toggleCard', function(toggle)
     local xPlayer = QBCore.Functions.GetPlayer(src)
 
     while xPlayer == nil do Wait(0) end
```
Comment (BerkieBb) [inline server/main.lua:312]: Same for this one as the one above this

File: server/main.lua [hunk 6]
```diff
@@ -278,23 +310,36 @@ RegisterNetEvent('qb-banking:toggleCard', function(toggle)
...
+    else
+        TriggerClientEvent('QBCore:Notify', src, Lang:t('error.too_far_away'), "error")
+        TriggerEvent('qb-log:server:CreateLog', 'banking', 'Banking', "red", "**"..GetPlayerName(xPlayer.PlayerData.source) .. " (citizenid: "..xPlayer.PlayerData.citizenid.." | id: "..xPlayer.PlayerData.source..")** has attempted to lock/unlock a debit card whilst not at a banking location, Nearest bank = "..auth.nearestDistance)
+    end
 end)
 
 RegisterNetEvent('qb-banking:doQuickWithdraw', function(amount, branch)
     local src = source
     local xPlayer = QBCore.Functions.GetPlayer(src)
     while xPlayer == nil do Wait(0) end
```
Comment (BerkieBb) [inline server/main.lua:326]: And for this, etc.

File: server/main.lua [hunk 7]
```diff
@@ -304,54 +349,79 @@ RegisterNetEvent('qb-banking:updatePin', function(pin)
...
-        TriggerEvent('qb-log:server:CreateLog', 'banking', 'Banking', 'lightgreen', "**"..GetPlayerName(xPlayer.PlayerData.source) .. " (citizenid: "..xPlayer.PlayerData.citizenid.." | id: "..xPlayer.PlayerData.source..")** made a savings deposit of $"..tostring(amount).." successfully..")
+    
+    local auth = validatePosition(GetEntityCoords(GetPlayerPed(src)))
+    if auth.isNearBank then
+        local currentBank = xPlayer.Functions.GetMoney('bank')
+        if tonumber(amount) <= currentBank then
+            local bank = xPlayer.Functions.RemoveMoney('bank', tonumber(amount))
+            local savings = savingsAccounts[xPlayer.PlayerData.citizenid].AddMoney(tonumber(amount), Lang:t('info.current_to_savings'))
+            while bank == nil do Wait(0) end
+            while savings == nil do Wait(0) end
```
Comment (BerkieBb) [inline server/main.lua:377]: Please don't use while loops here as the variables are not getting assigned later anytime, if it is nil it stays nil, so please make a check like this `if not bank then return end` you can add a message to that check as well if you want

File: server/main.lua [hunk 8]
```diff
@@ -304,54 +349,79 @@ RegisterNetEvent('qb-banking:updatePin', function(pin)
...
     local xPlayer = QBCore.Functions.GetPlayer(src)
-    local success = createSavingsAccount(xPlayer.PlayerData.citizenid)
-    repeat Wait(0) until success ~= nil
-    TriggerClientEvent('qb-banking:openBankScreen', src)
-    TriggerClientEvent('qb-banking:successAlert', src, Lang:t('success.opened_savings'))
-    TriggerEvent('qb-log:server:CreateLog', 'banking', 'Banking', "lightgreen", "**"..GetPlayerName(xPlayer.PlayerData.source) .. " (citizenid: "..xPlayer.PlayerData.citizenid.." | id: "..xPlayer.PlayerData.source..")** opened a savings account")
+    local auth = validatePosition(GetEntityCoords(GetPlayerPed(src)))
+    if auth.isNearBank then
+        local success = createSavingsAccount(xPlayer.PlayerData.citizenid)
+        repeat Wait(0) until success ~= nil
```
Comment (BerkieBb) [inline server/main.lua:417]: Also no repeat statement needed here, just an if statemenf as said above for the while loops

**Review by BerkieBb — CHANGES_REQUESTED** [review]

**Review by ItsANoBrainer — APPROVED** [review]

**Pull request closed without merging**